### ISSUE TITLE
Ibus has stopped working in specific apps, deletes preceding character (arch linux, i3)

### ISSUE THREAD

**Issue by ashprice** [ibus/ibus#2299]
Apologies for the non-descript title.

Which distribution and version?: 
Arch Linux 5.4.95-1-lts

Which desktop environment and version?: 
i3-gaps 4.19.1-1

Which session type?: 
X11

Which application and version?: 
firefox 85.0.1-1, discord 0.0.13-1, st-git 0.8.3.r19.g9ba7ecf-1, emacs-git 28.0.50.145250-1, libreoffice-still 7.0.4-1 etc. Will test any applications you want me to. Ibus _works_ in anki 2.1.35-3, not sure if it works in anything else. And there there is an issue, too - it only works for the first 3/4 of a line in flashcard entry. :D

IBus version?: 
IBus 1.5.23

Issue description:
Ibus only works in Anki. I can switch inputs fine, but when I go to type anything that _isn't_ English (which works, the English keyboard being the English (UK, extended, Windows) layout), like ibus-mozc, SIL's Eurolatin, the m17n Ancient Greek (polytonic support), the m17n Hindi inscript layout, SIL's keyman IPA unicode 6.2 layout, the m17n Persian and Russian layouts, etc. Basically, all the layouts I have tried/use don't work. Keypresses are registered - what happens is the _preceding_ character is deleted. Given a string _abcdefg_ every keypress will delete 1 character -> _abcdef_ -> _abcde_ etc.

Steps to reproduce:
1. Setup Arch Linux and i3, install ibus and any testing layouts.
2. Add the following lines to `/etc/profile`:
```
export GTK_IM_MODULE=xim
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=ibus
export XIM_PROGRAM=/usr/bin/ibus-daemon -drx
```
(Actually I just noticed, I couldn't copy and paste this into firefox how I usually do...)
3, Run `ibus-daemon -drx` in a terminal
4. Try entering text in the apps given above. 

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
(Run `ibus exit` and `ibus-daemon --xim &`)
Yes. Tested with both killall ibus-daemon/ibus exit and both ibus-daemon --xim and ibus-daemon -drx.

Do you see any errors when you run ibus-daemon with the verbose option?:
(Run `ibus-daemon --xim --verbose &` and look at the output when you encounter your problem.)
A GTK warning, Locale not supported by C library. Says it uses the fallback C locale. Other than that, a few errors about authors and sites in the jsons, which I assume are unproblematic. At the end there is the following error:
`ibus_bus_call_sync: org.freedesktop.DBus.Properties.Get: GDBus.Error:org.freedesktop.DBus.Error.Failed: No global engine.`
Further down there is this error:`
GLib-CRITICAL **: 10:12:28:047: g_queue_forsearch: assertion 'func != NULL' failed`
Can you reproduce your problem with a new user account instead of the current your account? (yes / no):
Will test and update later. 

Thank you!

**Comment by ashprice** [comment]
An update. Yes, I can replicate on new users. 

Setting this in /etc/profile:

```
export GTK_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=ibus
export XIM_PROGRAM=/usr/bin/ibus-daemon -drx
```

Fixes issues with firefox and discord. Originally that xim was actually in place to fix an issue with browsers, so I am unsure the cause for the switcharound. 

However, emacs, st still do not work. I did try changing the XMODIFERS to =ibus and =@im without luck.

**Comment by ashprice** [comment]
To clarify, this doesn't work either:
```
export GTK_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=xim
export XIM_PROGRAM=/usr/bin/ibus-daemon -drx
```
Curiously, if I set
```
export XMODIFIERS="@im=ibus" 
export XIM=ibus export XIM_ARGS="-d" 
export XIM_PROGRAM_SETS_ITSELF_AS_DAEMON=yes 
export DEPENDS="ibus" 
export XIM_PROGRAM=/usr/bin/ibus-daemon 
export GTK_IM_MODULE=ibus 
export QT_IM_MODULE=ibus 
```
(being sure to use sudo!) and log out and log in, /etc/profile resets to whatever was previously set. (ie. unlike the section above, where both GTK_IM and QT_IM are ibus, it will set xim.

**Comment by ashprice** [comment]
Solved the issue. My locales had been overwritten, presumably by the recent glibc updates. I noticed because font support had broken in some applications written in C. Adding en_GB.UTF-8 fixed the issues with Latin scripts and with input in non-C apps, adding ja_JP.UTF-8 fixed the issues with CJK scripts.

This is my current /etc/profile:
```
export GTK_IM_MODULE=ibus
export XMODIFIERS=@im=ibus
export QT_IM_MODULE=ibus
export XIM_PROGRAM=/urs/bin/ibus-daemon -drx
```

Some problems that didn't exist before exist now. First, ibus-mozc-ut decided to stop working for some reason. I had to fallback on ibus-mozc. Not the biggest issue, though I've already had to add quite a few lexical items to the IME's dictionary. Second, in some programs, or rather in certain contexts in certain programs, input can prove a bit... off. For example, in typing ð, \ð will appear, sometimes characters will be highlighted, sometimes backspaces will not work, sometimes writing will overwrite the following character. That specific behaviour happens in places like the google search bar and discord's messaging space. Here, on github, I can't type using my preferred Latin based input method (a keyman based layout), only the default one, the m17n ones, etc. and other scripts like 日本語, aɪː'pʰiːeiː, فارسی, русский, etc., but writing, say ǫ, å, ä, ö, ß, æ, ş with the preferred input doesn't work. There are also issues with backspace; sometimes characters will delete, sometimes it will skip over one, highlighting the one it skipped over, then deleting the one preceding that if pressed again (but leaving the skipped over character there), etc. I will try setting the QT_IM_MODULE and GTK_IM_MODULE variables differently, and to reinstall/reconfigure. 

In any case, the issue that prompted me to start the issue have been solved.
